src/object/cli_shard.c

```diff
@@ -725,71 +728,118 @@ dc_shard_update_size(struct rw_cb_args *rw_args, int fetch_rc)
 				iod->iod_size = sizes[i];
 			}
-		} else {
-			/* single-value, trust the size replied from first shard or parity shard,
-			 * because if overwrite those shards must be updated.
-			 */
-			if ((orw->orw_oid.id_shard % obj_ec_tgt_nr(oca)) ==
-			     obj_ec_singv_small_idx(oca, iod) ||
-			    is_ec_parity_shard(orw->orw_oid.id_shard, oca)) {
-				if (uiod->iod_size != 0 && uiod->iod_size < sizes[i] &&
-				    fetch_rc == 0) {
-					rc = -DER_REC2BIG;
-					D_ERROR(DF_UOID" original iod_size "DF_U64
-						", real size "DF_U64", "DF_RC"\n",
-						DP_UOID(orw->orw_oid),
-						iod->iod_size, sizes[i], DP_RC(rc));
-					iod->iod_size = sizes[i];
-					uiod->iod_size = sizes[i];
-					goto unlock;
-				}
+			goto unlock;
+		}
+
+		/* single-value, trust the size replied from first shard or parity shard,
+		 * because if overwrite those shards must be updated.
+		 */
+		if ((orw->orw_oid.id_shard % obj_ec_tgt_nr(oca)) ==
+		     obj_ec_singv_small_idx(oca, iod) ||
+		    is_ec_parity_shard(orw->orw_oid.id_shard, oca)) {
+			if (uiod->iod_size != 0 && uiod->iod_size < sizes[i] && fetch_rc == 0) {
+				rec2big = true;
+				rc = -DER_REC2BIG;
+				D_ERROR(DF_UOID" original iod_size "DF_U64", real size "DF_U64
+					", "DF_RC"\n", DP_UOID(orw->orw_oid),
+					iod->iod_size, sizes[i], DP_RC(rc));
 				iod->iod_size = sizes[i];
 				uiod->iod_size = sizes[i];
-				if (reasb_req->orr_size_set[i] == 0) {
-					reasb_req->orr_size_set[i] = sizes[i];
-				} else if (sizes[i] != reasb_req->orr_size_set[i]) {
-					conflict = dc_shard_singv_size_conflict(oca,
-							reasb_req->orr_size_set[i], sizes[i]);
-				}
-			} else if (sizes[i] != 0) {
-				if (iod->iod_size == 0)
-					iod->iod_size = sizes[i];
-				if (reasb_req->orr_size_set[i] == 0) {
-					reasb_req->orr_size_set[i] = sizes[i];
-				} else if (sizes[i] != reasb_req->orr_size_set[i]) {
-					conflict = dc_shard_singv_size_conflict(oca,
-							sizes[i], reasb_req->orr_size_set[i]);
+				goto unlock;
+			}
+
+			iod->iod_size = sizes[i];
+			uiod->iod_size = sizes[i];
+			if (fetch_stat->sfs_size == 0) {
+				fetch_stat->sfs_size = sizes[i];
+			} else if (fetch_stat->sfs_size != sizes[i]) {
+				rc = -DER_IO;
+				D_ERROR(DF_UOID" size mismatch "DF_U64" != "DF_U64", "DF_RC"\n",
+					DP_UOID(orw->orw_oid), fetch_stat->sfs_size, sizes[i],
+					DP_RC(rc));
+				goto unlock;
+			}
+
+			if (fetch_stat->sfs_size_other != 0 && fetch_rc == 0 &&
+			    fetch_stat->sfs_rc_other == 0) {
+				conflict = dc_shard_singv_size_conflict(oca,
+						fetch_stat->sfs_size_other, fetch_stat->sfs_size);
+			}
+			rc = fetch_rc;
+			/* one case needs to ignore the DER_REC2BIG failure - long singv
+			 * overwritten by short singv and the short singv only store on one
+			 * data target (and parity targets), other cases should return
+			 * DER_REC2BIG.
+			 */
+			if (rc == 0 && fetch_stat->sfs_rc_other == -DER_REC2BIG &&
+			    !obj_ec_singv_one_tgt(fetch_stat->sfs_size, NULL, oca)) {
+				rc = -DER_REC2BIG;
```

Comment:
What's the purpose to set "rc" as -DER_REC2BIG? that will prevent "fetch_again = true" at line 830. If it is expected, do we need it after line 821?

Reply:
yes, it is expected. actually set "rc = -DER_REC2BIG" is not must needed, as REC2BIG is higher priority than FETCH_AGAIN when return error code.

"do we need it after line 821"
note that there possibly with multiple IODs, and may one IOD need not fetch again, and another IOD need fetch again. So I added some other check below.

Reply:
i removed the rc setting here to avoid confuse, thx